### PR TITLE
20180919 nrcan15

### PR DESCRIPTION
**Pull request by mmpagani**
This PR adds preliminary versions of the GMM for the 2015 version of the Canada model implemented in an analytical form. The tests at the moment use very large tolerances due to discrepancies between the computed GM values and the reference ones included in the tables published on http://www.seismotoolbox.ca/. Hence this version should be considered PRELIMINARY and UNTESTED. It is merged into master since it will help in better understanding the impact on hazard results of the differences observed between the analytical model (the current PR) and the tabular GMPEs originally used for the calculation of hazard.

### CODE REVIEW

File: openquake/hazardlib/gsim/can15/eastern.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    REQUIRES_RUPTURE_PARAMETERS = set(('rake', 'mag'))
+
+    # Distances to be excluded while checking this GMPE
+    DO_NOT_CHECK_DISTANCES = set(('rrup', 'rjb'))
+
+    def apply_correction_to_BC(self, mean, imt, dists):
+        """
+        """
+        if re.search('^SA', str(imt)):
+            cff = self.SITE_COEFFS[imt]
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/eastern.py]: You can just use

```python
if imt.period:
  ...
else:
 ...
```
an AttributeError will be raised if the IMT is not SA or PGA which should be good enough.

File: openquake/hazardlib/gsim/can15/sinter.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+from openquake.hazardlib.gsim.zhao_2006 import ZhaoEtAl2006SInter
+from openquake.hazardlib.gsim.atkinson_macias_2009 import AtkinsonMacias2009
+from openquake.hazardlib.gsim.abrahamson_2015 import AbrahamsonEtAl2015SInter
+from openquake.hazardlib.gsim.ghofrani_atkinson_2014 import \
+    GhofraniAtkinson2014
+
+
+class SInterCan15Mid(ZhaoEtAl2006SInter):
+    """
+    """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/sinter.py]: Missing docstring

File: openquake/hazardlib/gsim/can15/oceanic.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+from openquake.hazardlib.gsim.can15.western import get_sigma
+
+
+class OceanicCan15Mid(WesternCan15Mid):
+    """
+    Implements the GMPE for oceanic sources
+    """
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/oceanic.py]: If a docstring is not interesting, it is better to remove it than to have it empty

File: openquake/hazardlib/gsim/can15/oceanic.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        mean, stddevs = super().get_mean_and_stddevs(sites, rup, dists, imt,
+                                                     stddev_types)
+        stddevs = [np.ones(len(dists.rjb))*get_sigma(imt)]
+        return mean, stddevs
+
+
+class OceanicCan15Low(WesternCan15Mid):
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/oceanic.py]: Same comment on the docstring

File: openquake/hazardlib/gsim/can15/oceanic.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        delta = np.log(10.**(np.amin(tmp, axis=0)))
+        mean_adj = mean - delta
+        stddevs = [np.ones(len(dists.rjb))*get_sigma(imt)]
+        return mean_adj, stddevs
+
+
+class OceanicCan15Upp(WesternCan15Mid):
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/oceanic.py]: Idem

File: openquake/hazardlib/gsim/can15/sinter.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        delta = np.minimum((0.15-0.0007*dists.rrup), 0.35)
+        return delta
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        mean, stds = self._get_mean_and_stddevs(sites, rup, dists, imt,
+                                                stddev_types)
+        return mean, stds
+
+    def _get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/sinter.py]: Idem

File: openquake/hazardlib/gsim/can15/sslab.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    """
+
+    # Required distance is only repi since rrup and rjb are obtained from repi
+    REQUIRES_DISTANCES = set(('repi',))
+
+    # Distances to be excluded while checking this GMPE
+    DO_NOT_CHECK_DISTANCES = set(('rrup', 'rjb'))
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/sslab.py]: Idem

File: openquake/hazardlib/gsim/can15/sslab.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    3.000   0.079
+    5.000   0.040
+    10.00   0.000
+    """)
+
+
+class SSlabCan15Low(SSlabCan15Mid):
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/sslab.py]: Idem

File: openquake/hazardlib/gsim/can15/sslab.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        # Adams, 2013; page 992)
+        delta = np.log(10.**(0.15))
+        mean_adj = mean - delta
+        return mean_adj, stddevs
+
+
+class SSlabCan15Upp(SSlabCan15Mid):
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/sslab.py]: Idem

File: openquake/hazardlib/gsim/can15/western.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+def get_sigma(imt):
+    """
+    Return the value of the total sigma
+
+    :param float imt:
+        An :class:`openquake.hazardlib.imt.IMT` instance
+    :returns:
+        A float representing the total sigma value
+    """
+    if str(imt) == 'PGA' or imt.period < 0.2:
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/western.py]: You can just write `if imt.period < 0.2`.

File: openquake/hazardlib/gsim/can15/western.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+class WesternCan15Mid(BooreAtkinson2011):
+    """
+    Implements the Boore and Atkinson (2008) with adjustments proposed by
+    Boore and Atkinson (2011) and the modifications introduced for the
+    calculation of hazard for the fifth generation of Canada's hazard maps,
+    released in 2015.
+    """
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/western.py]: Empty docstring again

File: openquake/hazardlib/gsim/can15/western.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        mean, stddevs = super().get_mean_and_stddevs(sites, rup, dists, imt,
+                                                     stddev_types)
+        stddevs = [np.ones(len(dists.rjb))*get_sigma(imt)]
+        return mean, stddevs
+
+
+class WesternCan15Low(WesternCan15Mid):
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/western.py]: Idem

File: openquake/hazardlib/gsim/can15/western.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        delta = np.log(10.**(np.amin(tmp, axis=0)))
+        mean_adj = mean - delta
+        stddevs = [np.ones(len(dists.rjb))*get_sigma(imt)]
+        return mean_adj, stddevs
+
+
+class WesternCan15Upp(WesternCan15Mid):
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """ """
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/western.py]: Idem

File: openquake/hazardlib/tests/gsim/can15/eastern_test.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class EasternCan15MidTestCase(BaseGSIMTestCase):
+    GSIM_CLASS = EasternCan15Mid
+
+    def test_mean(self):
+        self.check('CAN15/GMPEt_ENA_med.csv',
+                   max_discrep_percentage=100.)
+    """
+    def test_std_total(self):
```
Comment (micheles) [inline openquake/hazardlib/tests/gsim/can15/eastern_test.py]: Is this commented code? If it should stay for future use there please say so in a comment.

File: openquake/hazardlib/gsim/can15/sinter.py (hunk 2 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/sinter.py]: Fixed

File: openquake/hazardlib/gsim/can15/oceanic.py (hunk 3 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/oceanic.py]: Fixed

File: openquake/hazardlib/gsim/can15/oceanic.py (hunk 4 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/oceanic.py]: Fixed

File: openquake/hazardlib/gsim/can15/oceanic.py (hunk 5 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/oceanic.py]: idem

File: openquake/hazardlib/gsim/can15/sinter.py (hunk 6 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/sinter.py]: idem

File: openquake/hazardlib/gsim/can15/sslab.py (hunk 7 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/sslab.py]: Fixed

File: openquake/hazardlib/gsim/can15/sslab.py (hunk 8 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/sslab.py]: Fixed

File: openquake/hazardlib/gsim/can15/sslab.py (hunk 9 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/sslab.py]: Fixed

File: openquake/hazardlib/gsim/can15/western.py (hunk 11 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/western.py]: Fixed

File: openquake/hazardlib/gsim/can15/western.py (hunk 13 above)
Comment (mmpagani) [reply openquake/hazardlib/gsim/can15/western.py]: Fixed

File: openquake/hazardlib/tests/gsim/utils.py [hunk 15]
```diff
@@ -29,21 +29,24 @@ class BaseGSIMTestCase(unittest.TestCase):
...
-        self.assertTrue(
-            numpy.all(ctxs),
-            msg='Contexts objects have been changed by method '
-                'get_mean_and_stddevs')
+        if not hasattr(gsim, 'DO_NOT_CHECK_DISTANCES') and not admit_changes:
+            self.assertTrue(
+                numpy.all(ctxs),
+                msg='Contexts objects have been changed by method '
+                    'get_mean_and_stddevs')
         if errors:
```
[the remark this answers is not in the document]
Comment (mmpagani) [reply openquake/hazardlib/tests/gsim/utils.py]: It works

File: openquake/hazardlib/gsim/can15/eastern.py [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+    """
+    Implements the hybrid GMPE used to compute hazard in the Eastern part of
+    Canada.
+
+    The GMPEs used are:
+
+    - Pezeshk et al. (2011) - For this GMPE we scale the ground motion from
+    hard rock to B/C using the correction proposed in Atkinson and Adams (2013)
+    Table 2 page 994. As the distance metric used is Rrup we compute and
+    equivalent Rrup distance from Repi using the equations in Appendix A (page
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/eastern.py]: s/and/an/

File: openquake/hazardlib/gsim/can15/eastern.py [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+    31) of Atkinson (2012).
+
+    - Atkinson (2008) as revised in Atkinson and Boore (2011). As the distance
+    metric used is Rjb we compute and equivalent Rjb distance from Repi
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/eastern.py]: s/and/an

File: openquake/hazardlib/gsim/can15/eastern.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+    using the equations in Appendix A (page 31) of Atkinson (2012).
+
+    - Atkinson and Boore (2006) as revised in Atkinson and Boore (2011). As
+    the distance metric used is Rjb we compute and equivalent Rjb distance
```
Comment (micheles) [inline openquake/hazardlib/gsim/can15/eastern.py]: s/and/an

**Pull request merged**